### PR TITLE
[IMP] stock: improve the alert system

### PR DESCRIPTION
**Pull request by std-odoo**
Sale stock
==========
Split the widget `qty_at_date_widget` so we can use it in other models/

MRP - Stock - Sale stock
===================
Modify the alert system. Instead of writing exception in the chatter, we display a popover widget.

Stock move
=========
Add a field on `stock.move`, `delay_alert_date`. This field is False is the stock move will not be late, otherwise, this field contains the minimal date that the stock move should have.

Task #1970595

### CODE REVIEW

File: addons/mrp/models/mrp_production.py [hunk 1]
```diff
@@ -221,6 +224,59 @@ def _get_default_date_planned_finished(self):
...
+    def action_reschedule(self):
+        if self.move_raw_ids:
+            self.move_raw_ids.action_group_reschedule()
+            self.date_planned_finished = self.date_planned_start + timedelta(hours=1)
+
+    @api.depends('move_raw_ids')
+    def _compute_previous_mrp_production_id(self):
+        for mrp_production in self:
+            if mrp_production.move_raw_ids:
+                prev_stock_moves = self.env['mrp.production'].search([('move_dest_ids', 'in', mrp_production.move_raw_ids.ids)])
```
Comment (Whenrow) [inline addons/mrp/models/mrp_production.py]: move_dest_ids c'est sur les stock.move et là tu cherche parmis des mrp.production

File: addons/mrp/tests/test_procurement.py [hunk 2]
```diff
@@ -183,17 +183,20 @@ def test_date_propagation(self):
         self.assertEqual(mo_form.product_qty, 10, 'the quantity to produce is not good relative to the move')
 
         mo = mo_form.save()
-
         # Confirming mo create finished move
         move_orig = self.env['stock.move'].search([
             ('move_dest_ids', 'in', move_dest.ids)
         ], limit=1)
+        move_orig.write({'propagate_date': True, 'propagate_date_minimum_delta': 1})
```
Comment (sle-odoo) [inline addons/mrp/tests/test_procurement.py]: the manufacture rule already got propagate date true en delta 1 isn't it? why this change?

File: addons/mrp/tests/test_procurement.py [hunk 3]
```diff
@@ -183,17 +183,20 @@ def test_date_propagation(self):
...
 
         self.assertEqual(len(move_orig), 1, 'the move orig is not created')
         self.assertEqual(move_orig.product_qty, 10, 'the quantity to produce is not good relative to the move')
 
         move_dest_scheduled_date = move_dest.date_expected
         mo.date_planned_start += timedelta(days=5)
+
+        self.assertAlmostEqual(move_orig.date_expected, move_dest_scheduled_date + timedelta(days=4, hours=23), delta=timedelta(seconds=1))
```
Comment (sle-odoo) [inline addons/mrp/tests/test_procurement.py]: add a comment: check that writing on the MO planned date do change the move finished date_expected

File: addons/stock/models/stock_move.py [hunk 4]
```diff
@@ -276,7 +278,7 @@ def _compute_product_qty(self):
                 move.product_uom_qty, move.product_id.uom_id, rounding_method=rounding_method)
 
     def _get_move_lines(self):
-        """ This will return the move lines to consider when applying _quantity_done_compute on a stock.move. 
+        """ This will return the move lines to consider when applying _quantity_done_compute on a stock.move.
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: remove

File: addons/stock/models/stock_move.py [hunk 5]
```diff
@@ -339,7 +341,7 @@ def _compute_reserved_availability(self):
         and is represented by the aggregated `product_qty` on the linked move lines. If the move
         is force assigned, the value will be 0.
         """
-        result = {data['move_id'][0]: data['product_qty'] for data in 
+        result = {data['move_id'][0]: data['product_qty'] for data in
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: remove

File: addons/stock/models/stock_move.py [hunk 6]
```diff
@@ -464,18 +466,8 @@ def write(self, vals):
...
-                if not move_dest_ids:
-                    continue
-                delta_days = (new_date - move.date_expected).total_seconds() / 86400
-                if not move.propagate_date or abs(delta_days) < move.propagate_date_minimum_delta:
-                    move_dest_ids._delay_alert_log_activity('manual', move)
-                    continue
-                for move_dest in move_dest_ids:
-                    move_dest.date_expected += relativedelta.relativedelta(days=delta_days)
-                move_dest_ids._delay_alert_log_activity('auto', move)
+                move._delay_alert_log_activity(vals.get(propagated_date_field))
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: before the method was in two part
1) reschedule
2) delay alert

now you do the rescheduling in the delay alert method?
also i don't think this should be common: the rescheduling work on the next move, the delay alert has a new special smart logic which is more complex

File: addons/mrp/models/mrp_production.py [hunk 7]
```diff
@@ -221,6 +224,56 @@ def _get_default_date_planned_finished(self):
...
+    @api.depends('move_raw_ids.date_expected')
+    def _compute_date_planned_start(self):
+        for mrp_production in self:
+            if mrp_production.move_raw_ids:
+                mrp_production.date_planned_start = max(mrp_production.mapped('move_raw_ids.date_expected'))
+            else:
+                mrp_production.date_planned_start = fields.Datetime.now()
+
+    def _set_date_planned_start(self):
+        for mrp_production in self:
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: split in a proper commit the removal of the onchange for the compute with an inverse please

File: addons/stock/models/stock_move.py [hunk 8]
```diff
@@ -179,11 +179,48 @@ def _default_group_id(self):
...
+
     @api.onchange('product_id', 'picking_type_id')
     def onchange_product(self):
         if self.product_id:
             self.description_picking = self.product_id._get_description(self.picking_type_id)
 
+    @api.depends('date_expected', 'move_orig_ids.date_expected', 'date', 'state')
+    def _compute_will_be_late(self):
+        for stock_move in self:
+            stock_move.will_be_late = stock_move.delay_alert and bool(stock_move._get_delay_time())
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: i don't like it

i would prefer a normal stored boolean named `late` updated during write

File: addons/mrp/models/mrp_production.py [hunk 9]
```diff
@@ -114,6 +114,7 @@ def _get_default_date_planned_finished(self):
         help="Work orders will be planned based on the availability of the work centers\
               starting from this date. If empty, the work orders will be planned as soon as possible.",
     )
+    delay_alert_date = fields.Datetime('Delay Alert Date', default=False, compute='_compute_delay_alert_date')
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: default not needed

File: addons/stock/models/stock_move.py [hunk 10]
```diff
@@ -545,6 +537,63 @@ def _delay_alert_log_activity(self, mode, move_orig):
...
+        next_moves_dates += next_done_moves.mapped('date')
+        next_moves_dates += next_nondone_moves.mapped('date_expected')
+        if next_moves_dates:
+            next_moves_date = min(next_moves_dates)
+            if new_date > next_moves_date:
+                late_moves |= next_nondone_moves
+            else:
+                ontime_moves |= next_nondone_moves
+
+            late_moves.write({'delay_alert_date': new_date})
```
Comment (amoyaux) [inline addons/stock/models/stock_move.py]: The logic is maybe to simple. Transport delay are totally ignored. I don't know maybe check with usability team if it's okay.

File: addons/stock/models/stock_move.py [hunk 11]
```diff
@@ -545,6 +537,63 @@ def _delay_alert_log_activity(self, mode, move_orig):
...
+        previous_moves_dates += previous_nondone_moves.mapped('date_expected')
+        if previous_moves_dates:
+            previous_moves_date = max(previous_moves_dates)
+            if new_date < previous_moves_date:
+                late_moves |= self
+            else:
+                ontime_moves |= self
+
+            late_moves.write({'delay_alert_date': previous_moves_date})
+            ontime_moves.write({'delay_alert_date': False})
```
Comment (amoyaux) [inline addons/stock/models/stock_move.py]:
``` python3
            if new_date < previous_moves_date:
                self.write({'delay_alert_date': previous_moves_date})
            else:
                self.write({'delay_alert_date': False})
```

File: addons/stock/tests/test_move2.py [hunk 12]
```diff
@@ -2531,3 +2531,83 @@ def test_delay_alert_2(self):
...
+        self.assertFalse(pick.delay_alert_date)
+        self.assertFalse(pack.delay_alert_date)
+        self.assertFalse(ship.delay_alert_date)
+
+        # move the pack after the ship
+        # pick - ship - pack
+        pack.date_expected += timedelta(days=2)
+        self.assertFalse(pick.delay_alert_date)
+        self.assertFalse(pack.delay_alert_date)
+        self.assertTrue(ship.delay_alert_date)
```
Comment (amoyaux) [inline addons/stock/tests/test_move2.py]: Could assert the date when it's True. So it would both check if it's set and if the correct date is set

File: addons/stock/models/stock_picking.py [hunk 13]
```diff
@@ -377,6 +380,12 @@ def _compute_has_tracking(self):
         for picking in self:
             picking.has_tracking = any(m.has_tracking != 'none' for m in picking.move_lines)
 
+    @api.depends('move_lines.delay_alert_date')
+    def _compute_delay_alert_date(self):
+        for picking in self:
+            sorted_move_lines = picking.move_lines.filtered(lambda m: m.delay_alert_date).sorted(lambda m: m.delay_alert_date)
+            picking.delay_alert_date = sorted_move_lines[-1].delay_alert_date if sorted_move_lines else False
```
Comment (amoyaux) [inline addons/stock/models/stock_picking.py]: At this point, it would be interesting to check the date on the picking and the picking type delivery policy (ASAP, all at once). It's not true that the picking is always late if one move is late.

File: addons/stock/models/stock_move.py (hunk 11 above)
Comment (std-odoo) [reply addons/stock/models/stock_move.py]: @amoyaux with this, we will perform multiple write, instead of just 2 write, is it less performant ?

File: addons/stock/models/stock_picking.py (hunk 13 above)
Comment (std-odoo) [reply addons/stock/models/stock_picking.py]: @amoyaux the date of the moves will always modify `scheduled_date` on the stock picking, so it can be interesting to show the information whatever the delivery policy

what do you think @sle-odoo ?

**Comment by amoyaux** [comment]
Bug: alert_delay is not propagated to the PO receipt. Add it on:
- po_line field
- _prepare_purchase_order_line_from_procurement
- _prepare_stock_moves

Check propage_cancel as example

File: addons/stock/tests/test_move2.py [hunk 14]
```diff
@@ -2043,7 +2043,9 @@ def test_pickings_transfer_with_different_uom_and_back_orders(self):
         self.assertEqual(len(back_order_in), 1.00, 'There should be one back order created')
         self.assertEqual(back_order_in.move_lines.product_qty, 91640.00, 'There should be one back order created')
 
+from odoo.tests.common import tagged
 
+@tagged('blabla')
```
Comment (amoyaux) [inline addons/stock/tests/test_move2.py]: tag blabla

File: addons/mrp/models/mrp_production.py [hunk 15]
```diff
@@ -231,6 +234,25 @@ def _set_date_planned_start(self):
...
+        for production in self:
+            late_moves = production.move_raw_ids.filtered(lambda m: m.delay_alert_date).sorted(lambda m: m.delay_alert_date)
+            production.delay_alert_date = late_moves[-1].delay_alert_date if late_moves else False
+
+    def _compute_json_popover(self):
+        for production in self:
+            production.json_popover = json.dumps({
+                'popoverTemplate': 'mrp.ProductionPopover',
+                'model': 'mrp.production',
+                'delay_alert_date': production.delay_alert_date.strftime('%d/%m/%Y %H:%M:%S') if production.delay_alert_date else False,
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: to_string

File: addons/purchase_stock/models/purchase.py [hunk 16]
```diff
@@ -317,12 +317,11 @@ def _create_or_update_picking(self):
...
-                move_vals = line._prepare_stock_moves(picking)
-                for move_val in move_vals:
-                    self.env['stock.move']\
-                        .create(move_val)\
-                        ._action_confirm()\
-                        ._action_assign()
+
+                moves = line._create_stock_moves(picking)
+                for move in moves:
+                    move._delay_alert_check()
```
Comment (amoyaux) [inline addons/purchase_stock/models/purchase.py]: rem

File: addons/stock/models/stock_picking.py (hunk 13 above)
Comment (sle-odoo) [reply addons/stock/models/stock_picking.py]: we ignore the asap/all at once concept on these alerts at the moment

File: addons/mrp/models/mrp_production.py [hunk 17]
```diff
@@ -1057,3 +1079,8 @@ def _render_note_exception_quantity_mo(rendering_context):
             return self.env.ref('mrp.exception_on_mo').render(values=values)
 
         self.env['stock.picking']._log_activity(_render_note_exception_quantity_mo, documents)
+
+    @api.model
+    def _search_delay_alert_date(self, operator, value):
+        late_stock_moves = self.env['stock.move'].search([('delay_alert_date', '!=', False)])
+        return ['|', ('move_raw_ids', 'in', late_stock_moves.ids), ('move_finished_ids', 'in', late_stock_moves.ids)]
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: if you don't use operator and value here when using this field in a custom filter things will be weird: always the same result whatever the filter's value

File: addons/stock/models/stock_picking.py [hunk 18]
```diff
@@ -1265,3 +1287,8 @@ def _attach_sign(self):
             body=message,
         )
         return True
+
+    @api.model
+    def _search_delay_alert_date(self, operator, value):
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: search method should be near the compute

File: addons/stock/models/stock_picking.py [hunk 19]
```diff
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
+import json
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: put it next to "import time"  or move import time after this one, i don't remember the pep8 rule

File: addons/stock/models/stock_picking.py [hunk 20]
```diff
@@ -403,6 +412,19 @@ def _compute_show_lots_text(self):
             else:
                 picking.show_lots_text = False
 
+    def _compute_json_popover(self):
+        for picking in self:
+            picking.json_popover = json.dumps({
+                'popoverTemplate': 'stock.PopoverStockPicking',
+                'model': 'stock.picking',
+                'delay_alert_date': picking.delay_alert_date.strftime('%d/%m/%Y %H:%M:%S') if picking.delay_alert_date else False,
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: mmh mmh

File: addons/stock/models/stock_picking.py [hunk 21]
```diff
@@ -403,6 +412,19 @@ def _compute_show_lots_text(self):
...
+                'late_elements': [{
+                        'id': late_move.id,
+                        'name': late_move.display_name,
+                    } for late_move in picking.move_lines.filtered(lambda m: m.delay_alert_date).move_orig_ids.mapped('picking_id')
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: useless mapped

File: addons/stock/static/src/js/widget_popover_rescheduling.js [hunk 22]
```diff
@@ -0,0 +1,38 @@
+odoo.define('stock.PopoverStockPicking', function (require) {
+"use strict";
+
+var core = require('web.core');
+
+var PopoverWidget = require('stock.popover_widget');
```
Comment (sle-odoo) [inline addons/stock/static/src/js/widget_popover_rescheduling.js]: var PopoverWidgetField = require('stock.popover_widget');

File: addons/stock/static/src/js/widget_popover_rescheduling.js [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+                    views: [[false, 'form']],
+                    target: 'current'
+                });
+            });
+        }
+    },
+
+});
+
+registry.add('popover_stock_rescheduling', PopoverStockPicking);
```
Comment (sle-odoo) [inline addons/stock/static/src/js/widget_popover_rescheduling.js]: popover_stock_rescheduling -> stock_picking_popover (ryv introduced mrp_workorder_popover)

File: addons/stock/static/src/js/widget_popover_rescheduling.js [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+                    res_model: ev.currentTarget.getAttribute('element-model'),
+                    res_id: parseInt(ev.currentTarget.getAttribute('element-id')),
+                    views: [[false, 'form']],
+                    target: 'current'
+                });
+            });
+        }
+    },
+
+});
```
Comment (sle-odoo) [inline addons/stock/static/src/js/widget_popover_rescheduling.js]: maybe rename this file stock_picking_popover.js (ryv introduct mrp_workorder_popover.js)

File: addons/stock/views/stock_picking_views.xml [hunk 25]
```diff
@@ -294,7 +298,12 @@
                             <field name="backorder_id" attrs="{'invisible': [('backorder_id','=',False)]}"/>
                         </group>
                         <group>
+                            <label for="scheduled_date"/>
+                            <div class="o_row">
                             <field name="scheduled_date" attrs="{'required': [('id', '!=', False)]}"/>
+                                <field name="delay_alert_date" invisible="1"/>
+                                <field name="json_popover" string=" " widget="popover_stock_rescheduling" attrs="{'invisible': [('delay_alert_date', '=', False)]}"/>
```
Comment (sle-odoo) [inline addons/stock/views/stock_picking_views.xml]: why the string?

File: addons/mrp/models/mrp_production.py [hunk 26]
```diff
@@ -231,6 +234,25 @@ def _set_date_planned_start(self):
     def _set_date_planned_finished(self):
         self.move_finished_ids.write({'date_expected': self.date_planned_finished})
 
+    @api.depends('move_raw_ids.delay_alert_date')
+    def _compute_delay_alert_date(self):
+        for production in self:
+            late_moves = production.move_raw_ids.filtered(lambda m: m.delay_alert_date).sorted(lambda m: m.delay_alert_date)
+            production.delay_alert_date = late_moves[-1].delay_alert_date if late_moves else False
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: read_group and min on delay_alert ?

File: addons/mrp/views/mrp_templates.xml [hunk 27]
```diff
@@ -23,7 +23,7 @@
     </template>
 
     <template id="exception_on_mo">
-        <div class="alert alert-warning" role="alert"> 
+        <div class="alert alert-warning" role="alert">
```
Comment (sle-odoo) [inline addons/mrp/views/mrp_templates.xml]: useless diff

File: addons/sale_stock/models/sale_order.py [hunk 28]
```diff
@@ -40,6 +41,8 @@ def _default_warehouse_id(self):
                                           "the order lines in case of Service products. In case of shipping, the shipping policy of "
                                           "the order will be taken into account to either use the minimum or maximum lead time of "
                                           "the order lines.")
+    json_popover = fields.Char('JSON data for the popover widget', compute='_compute_json_popover')
+    has_late_picking = fields.Boolean('Has late picking', compute='_compute_json_popover')
```
Comment (sle-odoo) [inline addons/sale_stock/models/sale_order.py]: has_late_picking  -> show_ json_popover

File: addons/stock/static/src/js/widget_popover_rescheduling.js [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+        this._super();
+        if (this.$popover) {
+            var self = this;
+            this.$popover.find('a').on('click', function (ev) {
+                ev.preventDefault();
+                ev.stopPropagation();
+                self.do_action({
+                    type: 'ir.actions.act_window',
+                    res_model: ev.currentTarget.getAttribute('element-model'),
+                    res_id: parseInt(ev.currentTarget.getAttribute('element-id')),
```
Comment (sle-odoo) [inline addons/stock/static/src/js/widget_popover_rescheduling.js]: parseint requires a radix

File: addons/mrp/models/mrp_production.py [hunk 30]
```diff
@@ -237,6 +235,26 @@ def _set_date_planned_start(self):
...
+    def _compute_json_popover(self):
+        for production in self:
+            production.json_popover = json.dumps({
+                'popoverTemplate': 'mrp.ProductionPopover',
+                'model': 'mrp.production',
+                'delay_alert_date': format_datetime(self.env, production.delay_alert_date, dt_format=False) if production.delay_alert_date else False,
+                'late_elements': [{
+                        'id': late_move.id,
+                        'name': late_move.display_name,
+                    } for late_move in production.move_raw_ids.filtered(lambda m: m.delay_alert_date).move_orig_ids.production_id
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: this only handles child mo late for the parent mo
you also have
- po late for the components of an mo
- pbm late for its mo

File: addons/stock/models/stock_move.py [hunk 31]
```diff
@@ -135,6 +135,7 @@ def _default_group_id(self):
     propagate_date_minimum_delta = fields.Integer(string='Reschedule if Higher Than',
         help='The change must be higher than this value to be propagated')
     delay_alert = fields.Boolean('Alert if Delay')
+    delay_alert_date = fields.Datetime('Delay Alert Date', help='False is no delay, else contains the date which should have the stock move')
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: help="Process at this date to be on time"

File: addons/stock/models/stock_move.py [hunk 32]
```diff
@@ -544,6 +536,53 @@ def _delay_alert_log_activity(self, mode, move_orig):
                 user_id=doc.user_id.id or SUPERUSER_ID
             )
 
+    def _delay_alert_check(self, new_date=None):
+        """Set an alert on late moves by using the `delay_alert_date` field.
+        The alert is always on the move that cannot be done because its preceding moves are late.
+        """
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]:
``` python
:param new_date: the new expected date, to set when calling this method before `create` or `write`
:param new_date: datetime, optional
```

File: addons/mrp/views/mrp_production_views.xml [hunk 33]
```diff
@@ -325,6 +323,7 @@
                     <separator/>
                     <filter string="Late" name="late" help="Production started late"
                         domain="['&amp;', ('date_deadline', '&lt;', current_date), ('state', '=', 'confirmed')]"/>
+                    <filter string="Issues" name="will_be_late" domain="[('delay_alert_date', '!=', False)]"/>
```
Comment (sle-odoo) [inline addons/mrp/views/mrp_production_views.xml]: Planning Issues

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff
